From: Rainer Gerhards <rgerhards@adiscon.com>
Date: Tue, 21 Jan 2020 17:57:42 +0100
Subject: imfile bugfix: misadressing and potential segfault

Commit 3f72e8c introduced an invalid memory allocation size. This lead to
too-short alloc and thus to overwrite of non-owned memory. That in turn
could lead to segfaults or other hard to find problems.

The issue was detected by our upgraded CI system. We did not receive
any problem reports in practice. Nevertheless, the problem is real and
people should update affected versions to patched ones.

The bug was present in scheduled stable release 8.1911.0 and 8.2001.0.

see also: https://github.com/rsyslog/rsyslog/issues/4120
see also: https://github.com/rsyslog/rsyslog/pull/4141

(cherry picked from commit aa809b61dcadcb41c18a13e79276849ae75d6780)
---
 plugins/imfile/imfile.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/plugins/imfile/imfile.c b/plugins/imfile/imfile.c
index 8eb6763..3084a82 100644
--- a/plugins/imfile/imfile.c
+++ b/plugins/imfile/imfile.c
@@ -1144,7 +1144,8 @@ fs_node_add(fs_node_t *const node,
 		if(!ustrcmp(chld->name, name)) {
 			DBGPRINTF("fs_node_add(%p, '%s') found '%s'\n", chld->node, toFind, name);
 			/* add new instance */
-			instanceConf_t **instarr_new = realloc(chld->instarr, sizeof(instanceConf_t*) * chld->ninst+1);
+			instanceConf_t **instarr_new = realloc(chld->instarr,
+							sizeof(instanceConf_t*) * (chld->ninst+1));
 			CHKmalloc(instarr_new);
 			chld->instarr = instarr_new;
 			chld->ninst++;
